bokehjs/src/lib/models/ranges/factor_range.ts

```diff
@@ -186,119 +239,99 @@ export class FactorRange extends Range {
+  factor(x: number): Factor | null {
+    for (const f of this.factors) {
+      const v = this.synthetic(f)
+      if (x >= (v-0.5) && x < (v+0.5)) {
+        return f
+      }
     }
-    return array
+    return null
+  }
+
+  protected _compute_bounds(inner_padding: number): [number, number] {
+    const interval = this.factors.length + inner_padding
+    const padding = (() => {
+      switch (this.range_padding_units) {
+        case "percent": {
+          return interval * this.range_padding / 2
+        }
+        case "absolute": {
+          return this.range_padding
+        }
+      }
+    })()
+    // XXX "0 - padding" is a workaround for an assertions bug
+    return [0 - padding, interval + padding]
   }
 
   protected _init(): void {
-    const {levels, mapping, tops, mids, inside_padding} = (() => {
-      if (every(this.factors, isString)) {
-        const factors = this.factors as string[]
-        const [mapping, inside_padding] = map_one_level(factors, this.factor_padding)
-        const tops = null
-        const mids = null
-        return {levels: 1, mapping, tops, mids, inside_padding}
-      } else if (every(this.factors, (x) => isArray(x) && x.length == 2 && isString(x[0]) && isString(x[1]))) {
-        const factors = this.factors as [string, string][]
-        const [mapping, inside_padding] = map_two_levels(factors, this.group_padding, this.factor_padding)
-        const tops = [...mapping.keys()]
-        const mids = null
-        return {levels: 2, mapping, tops, mids, inside_padding}
-      } else if (every(this.factors, (x) => isArray(x) && x.length == 3 && isString(x[0]) && isString(x[1]) && isString(x[2]))) {
-        const factors = this.factors as [string, string, string][]
-        const [mapping, inside_padding] = map_three_levels(factors, this.group_padding, this.subgroup_padding, this.factor_padding)
-        const tops = [...mapping.keys()]
-        const mids: [string, string][] = []
-        for (const [f0, L2] of mapping) {
-          for (const f1 of L2.mapping.keys()) {
-            mids.push([f0, f1])
-          }
+
+    const levels = compute_levels(this.factors)
+    const {mapping, tops, mids, inner_padding} = (() => {
+      switch (levels) {
+        case 1: {
+          const factors = this.factors as L1Factor[]
+          return map_one_level(factors, this.factor_padding)
+        }
+        case 2: {
+          const factors = this.factors as L2Factor[]
+          return map_two_levels(factors, this.group_padding, this.factor_padding)
+        }
+        case 3: {
+          const factors = this.factors as L3Factor[]
+          return map_three_levels(factors, this.group_padding, this.subgroup_padding, this.factor_padding)
         }
-        return {levels: 3, mapping, tops, mids, inside_padding}
-      } else {
-        unreachable()
       }
     })()
 
-    this._mapping = mapping
+    this.mapping = mapping
     this.tops = tops
     this.mids = mids
```

Comment:
Is there any fancy Typescript way to improve or get rid of this clunk assignments?

Reply:
This is the right way